### PR TITLE
Coalesce sources until compressed serialized bundles under API limit

### LINKED ISSUE #26264: [Bug]: Elevated serialized size for BigQueryTableSource causing IllegalArgumentException in WorkerCustomSources
### What happened?

BigQuery export read Pipeline worked for Beam<=2.42.0 but failing in newer versions. Error message:

```
Error message from worker: java.lang.IllegalArgumentException: Total size of the BoundedSource objects generated by split() operation is larger than the allowable limit. When splitting org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@536a6236 into bundles of 565244165161 bytes it generated 5000 BoundedSource objects with total serialized size of 23719241 bytes which is larger than the limit 20971520
```

From the logging, the serialized size for BoundedSource elevated:

Failed job (v2.45.0):
```
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@6dfae6d4 into bundles of estimated size 5018826027 bytes produced 5000 bundles, which have total serialized size 1176766077 bytes. As this is too large for the Google Cloud Dataflow API, retrying splitting once with increased desiredBundleSizeBytes 563_238_545_888 to reduce the number of splits.
Splitting with desiredBundleSizeBytes 563238545888 produced 5000 bundles with total serialized size 1176766077 bytes
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@6dfae6d4 into bundles of estimated size 563238545888 bytes produced 5000 bundles. Rebundling into 100 bundles.
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@6dfae6d4 produced 100 bundles with total serialized response size 23625229
```
succeeded job (v2.42.0)
```
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@3fe148e1 into bundles of estimated size 5018826027 bytes produced 5000 bundles, which have total serialized size 612350329 bytes. As this is too large for the Google Cloud Dataflow API, retrying splitting once with increased desiredBundleSizeBytes 293_090_798_266 to reduce the number of splits.
Splitting with desiredBundleSizeBytes 293090798266 produced 5000 bundles with total serialized size 612350329 bytes
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@3fe148e1 into bundles of estimated size 293090798266 bytes produced 5000 bundles. Rebundling into 100 bundles.
Splitting source org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@3fe148e1 produced 100 bundles with total serialized response size 12673589
```

----

## Problem on the runner code

In both succeeded job and failed job, the total serialized size remains unchanged  in terms of reducing the number of bundles:
https://github.com/apache/beam/blob/837733e862007cbeba0a57f682109d9debec9340/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java#L239

only the second adjustment worked and change the number of the bundle to 100: https://github.com/apache/beam/blob/837733e862007cbeba0a57f682109d9debec9340/runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java#L253

This cannot guarantee the serialize limit has satisfied. There must some problem in splitAndValidate preventing it from working as expected.

----

## Problem on the serialized size

It shows the total serialized size increased from `612350329` to `1176766077`, that is 92% increase.

Did a minimum example Pipeline (reading a simple schema BigQuery using export), checking the log output for serialized source output for BigQueryTableSource, the strange thing is it shows opposite: the serialized size becomes smaller in Beam 2.43.0 and onwards:
```
v2.41.0: org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@1ab4652b produced 1 bundles with total serialized response size 3487
v2.42.0: org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@7072c138 produced 1 bundles with total serialized response size 3491
v2.43.0: org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@3648c304 produced 1 bundles with total serialized response size 2399
v2.46.0: org.apache.beam.sdk.io.gcp.bigquery.BigQueryTableSource@634fa80a produced 1 bundles with total serialized response size 2439
```
Looks like the serialized size of the source is less controlled.

### Issue Priority

Priority: 1 (data loss / total loss of function)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [x] Component: Google Cloud Dataflow Runner

### PR DESCRIPTION
**Pull request by Abacn**
Fixes #26264

The original `limitNumberOfBundles` logic has problems that it always limits to 100 bundles if the split bundle is greater than this number and did not check the resulting response size. It could result in IllegalArgumentException immediately below.

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
Verified that the unit test fails in current master and suceeds with fix in.

Current master:
```
Apr 13, 2023 4:25:34 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
WARNING: Splitting source <unknown> into bundles of estimated size 67108864 bytes produced 200 bundles, which have
total serialized size 116265 bytes. As this is too large for the Google Cloud Dataflow API, retrying splitting once with
increased desiredBundleSizeBytes 1560482414 to reduce the number of splits.
Apr 13, 2023 4:25:34 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
INFO: Splitting with desiredBundleSizeBytes 1560482414 produced 200 bundles with total serialized size 116265 bytes
Apr 13, 2023 4:25:34 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
WARNING: Splitting source <unknown> into bundles of estimated size 1560482414 bytes produced 200 bundles. Rebundling into 100 bundles.
Apr 13, 2023 4:25:34 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
INFO: Splitting source <unknown> produced 100 bundles with total serialized response size 58815

Total size of the BoundedSource objects generated by split() operation is larger than the allowable limit. When splitting
<unknown> into bundles of 1560482414 bytes it generated 200 BoundedSource objects with total serialized size of 58815
bytes which is larger than the limit 10000. For more information, please check the corresponding FAQ entry at
https://cloud.google.com/dataflow/pipelines/troubleshooting-your-pipeline
java.lang.IllegalArgumentException: Total size of the BoundedSource objects generated by split() operation is larger than the
allowable limit. When splitting <unknown> into bundles of 1560482414 bytes it generated 200 BoundedSource objects with
total serialized size of 58815 bytes which is larger than the limit 10000. For more information, please check the
corresponding FAQ entry at https://cloud.google.com/dataflow/pipelines/troubleshooting-your-pipeline
	at org.apache.beam.runners.dataflow.worker.WorkerCustomSources.performSplitTyped(WorkerCustomSources.java:286)
	at org.apache.beam.runners.dataflow.worker.WorkerCustomSources.performSplitWithApiLimit(WorkerCustomSources.java:201)
	at org.apache.beam.runners.dataflow.worker.WorkerCustomSourcesTest.testSplittingProducedResponseUnderLimit(WorkerCustomSourcesTest.java:242)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.rules.ExpectedException$ExpectedExceptionStatement.evaluate(ExpectedException.java:258)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$2.run(TestWorker.java:176)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
```

With the fix:
```
Apr 13, 2023 4:27:05 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
WARNING: Splitting source <unknown> into bundles of estimated size 67108864 bytes produced 200 bundles, which have
total serialized size 116265 bytes. As this is too large for the Google Cloud Dataflow API, retrying splitting once with
increased desiredBundleSizeBytes 1560482414 to reduce the number of splits.
Apr 13, 2023 4:27:05 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
INFO: Splitting with desiredBundleSizeBytes 1560482414 produced 200 bundles with total serialized size 116265 bytes
Apr 13, 2023 4:27:05 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
WARNING: Re-bundle source <unknown> into bundles of estimated size 13330 bytes produced 16 bundles.
Apr 13, 2023 4:27:06 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
WARNING: Re-bundle source <unknown> into bundles of estimated size 9814 bytes produced 11 bundles.
Apr 13, 2023 4:27:06 PM org.apache.beam.runners.dataflow.worker.WorkerCustomSources performSplitTyped
INFO: Splitting source <unknown> produced 11 bundles with total serialized response size 9824
```

**Comment by Abacn** [comment]
R: @chamikaramj @kennknowles per file history

internal tracker: 277995484

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java [hunk 1]
```diff
@@ -247,18 +247,37 @@ private static <T> SourceOperationResponse performSplitTyped(
...
+
+      if (targetBundleSize <= 1) {
+        String message =
+            String.format(
+                "Unable to coalesce the sources into compressed serialized bundles to satisfy the "
+                    + "allowable limit when splitting %s. With %d bundles, total serialized size "
+                    + "of %d bytes is still larger than the limit %d. For more information, please "
+                    + "check the corresponding FAQ entry at "
+                    + "https://cloud.google.com/dataflow/pipelines/troubleshooting-your-pipeline",
+                source, bundles.size(), serializedSize, apiByteLimit);
```
Comment (liferoad) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: Shall we suggest this workaround with `withTemplateCompatiblity`?

Comment (Abacn) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: This option is BigQueryIO.read only while the bug is generic to dataflow worker. With that option the read teansform will expand differently and likely avoided the problematic code path here (haven't verify)

Comment (liferoad) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: SG

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java [hunk 2]
```diff
@@ -247,18 +247,37 @@ private static <T> SourceOperationResponse performSplitTyped(
...
+      // bundle size constrained by numBundlesLimit
+      int targetBundleSizeBundleLimit = Math.min(numBundlesLimit, bundles.size() - 1);
+      int targetBundleSize = Math.min(targetBundleSizeApiLimit, targetBundleSizeBundleLimit);
+
+      if (targetBundleSize <= 1) {
+        String message =
+            String.format(
+                "Unable to coalesce the sources into compressed serialized bundles to satisfy the "
+                    + "allowable limit when splitting %s. With %d bundles, total serialized size "
+                    + "of %d bytes is still larger than the limit %d. For more information, please "
```
Comment (liferoad) [inline runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:266]: can we use this https://cloud.google.com/dataflow/docs/guides/common-errors#boundedsource-objects-splitintobundles? Please update the previous LOG.info with this link as well.

**Review by liferoad — APPROVED** [review]

**Comment by Abacn** [comment]
Run Java PreCommit

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java (hunk 2 above)
Comment (Abacn) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:266]: Good catch, this url is more accurate

**Review by liferoad — APPROVED** [review]

File: runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java (hunk 1 above)
Comment (chamikaramj) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: The workaround will be using Runner v2. That will convert sources into SDFs that do not run into this source split limit. We should suggest that here.

Comment (Abacn) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: Hi @chamikaramj is this code path only for Dataflow runner v1? If so we could document the recommendation of runner v2 also in https://cloud.google.com/dataflow/docs/guides/common-errors#boundedsource-objects-splitintobundles

Comment (chamikaramj) [reply runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WorkerCustomSources.java:269]: Yeah, this codepath (source split operation) is only get hit by Runner v1. +1 for documenting in the Website.

**Comment by chamikaramj** [comment]
LGTM

**Comment by Abacn** [comment]
Thanks for suggestions. The linked cloud documentation is going to be updated. Merging the fix for now.

**Pull request merged**